File: 2078-two-furthest-houses-with-different-colors/solution.py
class Solution:
    def maxDistance(self, colors: list[int]) -> int:
        n = len(colors)
        max_dist = 0
        
        # Scenario 1: Fix the first house, look for the furthest different color from the right
        for j in range(n - 1, 0, -1):
            if colors[j] != colors[0]:
                max_dist = max(max_dist, j)
                break
                
        # Scenario 2: Fix the last house, look for the furthest different color from the left
        for i in range(n - 1):
            if colors[i] != colors[n - 1]:
                max_dist = max(max_dist, (n - 1) - i)
                break
                
        return max_dist

File: 2078-two-furthest-houses-with-different-colors/solution.cpp
#include <vector>
#include <algorithm>

class Solution {
public:
    int maxDistance(std::vector<int>& colors) {
        int n = colors.size();
        int maxDist = 0;

        // Case 1: Compare index 0 with houses starting from the end
        for (int j = n - 1; j > 0; --j) {
            if (colors[j] != colors[0]) {
                maxDist = std::max(maxDist, j);
                break;
            }
        }

        // Case 2: Compare index n-1 with houses starting from the beginning
        for (int i = 0; i < n - 1; ++i) {
            if (colors[i] != colors[n - 1]) {
                maxDist = std::max(maxDist, (n - 1) - i);
                break;
            }
        }

        return maxDist;
    }
};